src/particles/ImpactXParticleContainer.cpp

```diff
@@ -121,4 +125,78 @@ namespace impactx
         return {xyz_min[0], xyz_min[1], xyz_min[2], xyz_max[0], xyz_max[1], xyz_max[2]};
     }
 
+    void
+    DepositCharge (ImpactXParticleContainer & pc, amrex::MultiFab* rho, int icomp) // ImpactXParticleContainer & pc)
+    {
+        // loop over refinement levels
+        int const nLevel = 1;
+        for (int lev = 0; lev < nLevel; ++lev) {
+            // get simulation geometry information
+            //const amrex::Geometry& gm = this->Geom(lev);
+            //const auto prob_lo = gm.ProbLo();
+
+            amrex::FArrayBox local_rho_fab;
+
+            // loop over all particle boxes
+            using ParIt = ImpactXParticleContainer::iterator;
+            for (ParIt pti(pc, lev); pti.isValid(); ++pti) {
+                const int np = pti.numParticles();
+                //const auto t_lev = pti.GetLevel();
+                //const auto index = pti.GetPairIndex();
+                // ...
+
+                // preparing access to particle data: SoA of Reals
+                auto &soa_real = pti.GetStructOfArrays().GetRealData();
+                //amrex::ParticleReal *const AMREX_RESTRICT wp = soa_real[RealSoA::w];
+                auto wp = soa_real[RealSoA::w];
+                int const * const AMREX_RESTRICT ion_lev = nullptr;
+
+                // DepositCharge(pti, wp, ion_lev, rho[lev].get(), icomp, 0, np, thread_num, lev, lev);
+
+                amrex::IntVect ref_ratio = amrex::IntVect(AMREX_D_DECL(1, 1, 1));
+                amrex::Box tilebox = pti.tilebox();
+
+                // TODO: physical lower corner of the current box
+                amrex::Geometry const &gm = pc.Geom(lev);
+                amrex::RealBox const grid_box{tilebox, gm.CellSize(), gm.ProbLo()};
+                amrex::Real const * const xyzmin_ptr = grid_box.lo();
+                std::array<amrex::Real, 3> const xyzmin = {xyzmin_ptr[0], xyzmin_ptr[1], xyzmin_ptr[2]};
+
+                // pointer to costs data
+                //amrex::LayoutData<amrex::Real>* costs = WarpX::getCosts(lev);
+                //amrex::Real* cost = costs ? &((*costs)[pti.index()]) : nullptr;
+                amrex::Real *cost = nullptr;
+
+                // used for MR when we want to deposit for a subset of the particles on the level in the
+                // current box; with offset, we start at a later particle index
+                const long offset = 0;
+                const long np_to_depose = np;
+                int depos_lev = lev;
+
```

Comment:
If we'd like to support OpenMP threading for CPU execution, then the `#pragma omp parallel` region should go here, including the `local_rho_fab` as a thread private variable.